### DISCUSSION TITLE
Label unreachable for Generator[..., ..., NoReturn]

### DISCUSSION THREAD

**Discussion by GreyElaina** [microsoft/pyright#6479] [Ideas]
```python
from __future__ import annotations
from typing import Any, Generator, NoReturn

a: Generator[int, Any, NoReturn] = ...

for i in a:
    reveal_type(i)

# below is unreachable due to a's ReturnType is NoReturn, which not exists for real logic.
# it means control flow will ends inside loop body (eg. return value / break) or end of iterating.
# it's useful...but first of all, is it wrong?

b = 1

reveal_type(b)
```

**Comment by erictraut** [comment]
The current behavior is intended. Only explicit function calls are evaluated reachability analysis. Evaluating all implicit magic methods for this purpose would be extremely expensive at analysis time. This needs to be done on a "hot path" during code flow analysis. It wouldn't be a good tradeoff to significantly slow down analysis time for all code just to support such an edge case as the one above.

**Reply by GreyElaina** [reply]
it makes sense.
